lock/src/main/java/org/infinispan/lock/api/impl/lock/ClusteredLockImpl.java

```diff
@@ -0,0 +1,369 @@
+            request.complete(true);
+            Boolean tryLockRealResult = request.join();
+            if (!tryLockRealResult) {
+               // Even if we complete true just before, the lock request can be completed false just before by the scheduler.
+               // This means that tryLock reached the max time waiting before the lock was actually acquired
+               // In this case, even if the lock was marked as acquired in the cache, it has to be released because the call expired.
+               // We have to unlock the lock if the requestor and the requestId match.
+               // Meanwhile another request for this owner might have locked it successfully and we don't want to unlock in that case
+               unlock(requestId, requestor);
+            }
+         } else if(!isScheduled){
+            // If the lock was not acquired, then schedule a complete false for the given timeout
+            isScheduled = true;
+            clusteredLockManager.schedule(() -> request.complete(false), time, unit);
+         }
+      }
+
+      @Override
+      public String toString() {
+         final StringBuilder sb = new StringBuilder("TryLockRequestHolder{");
+         sb.append(", requestId=").append(requestId);
+         sb.append(", requestor=").append(requestor);
+         sb.append(", time=").append(time);
+         sb.append(", unit=").append(unit);
+         sb.append(", completed=").append(request.isDone());
+         sb.append(", completedExceptionally=").append(request.isCompletedExceptionally());
+         sb.append('}');
+         return sb.toString();
+      }
+
+      public boolean hasTimeout() {
+         return time > 0;
+      }
+   }
+
+   @Listener(clustered = true)
+   class LockReleasedListener {
+
+      @CacheEntryModified
+      public void entryModified(CacheEntryModifiedEvent event) {
+         ClusteredLockValue value = (ClusteredLockValue) event.getValue();
+         if (value.getState() == ClusteredLockState.RELEASED) {
+            RequestHolder nextRequestor = null;
+            while (!pendingRequests.isEmpty() && (nextRequestor == null || nextRequestor.isDone()))
+               nextRequestor = pendingRequests.poll();
+
+            final RequestHolder requestor = nextRequestor;
+            clusteredLockManager.execute(() -> lock(requestor));
+         }
+      }
+
+      @CacheEntryRemoved
+      public void entryRemoved(CacheEntryRemovedEvent event) {
+         while (!pendingRequests.isEmpty()) {
+            pendingRequests.poll().handleLockResult(null, log.lockDeleted());
+         }
+      }
+   }
+
+   @Listener
+   class ClusterChangeListener {
+
+      @ViewChanged
+      public void viewChange(ViewChangedEvent event) {
+         log.trace("viewChange event has been fired");
+         List<Address> newMembers = event.getNewMembers();
+         List<Address> oldMembers = event.getOldMembers();
+
+         if(newMembers.size() > 1 || (oldMembers.size() == 2 && newMembers.size() == 1)) {
```

Comment:
It's because of the split brain vs one node is leaving. this part is a bit complicated now. 
The thing is that when I have a split brain or when a node just joins or reaves because of normal reasons, I get this event fired.
I just avoid in the case of an split brain where the partition is a single node, to fire unlocks for nothing. But if I have 2 nodes and one leaves, well, I want that other node to unlock even if the cluster is a single node now.
I can improve this, adding some more tests for that cases.

Reply:
You should not try to detect split brains ahead, that's a task for other components.
I was asking about two, because with more owners it should work well even with e.g. 2 nodes crashing/leaving simultaneously. The cache is replicated after all.